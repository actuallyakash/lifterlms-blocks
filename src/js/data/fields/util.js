/**
 * Redux data store utility functions
 *
 * @since [version]
 * @version [version]
 */

/**
 * Convert an array of user information fields to an object keyed by globally unique `name` attribute.
 *
 * @since [version]
 *
 * @param {Object[]} fields Array of field objects.
 * @return {Object} Object of field objects.
 */
export function fieldsArrayToObject( fields ) {
	return fields.reduce(
		( obj, currentField ) => ( {
			...obj,
			[ currentField.name ]: currentField,
		} ),
		{}
	);
}

/**
 * Convert an object of field objects into an array
 *
 * Useful for running array functions like filter.
 *
 * @since [version]
 *
 * @param {Object} fields Object of user information field objects.
 * @return {Object[]} Array of user information field objects.
 */
export function fieldsObjectToArray( fields ) {
	return Object.values( fields );
}
