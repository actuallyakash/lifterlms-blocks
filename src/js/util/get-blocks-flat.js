/**
 * Block List Flattening Utilities.
 *
 * Recursively loops through all nested blocks and returns
 * a flat array of every block in the layout.
 *
 * @since 1.6.0
 * @version 1.6.0
 */

// WP Deps.
const { select } = wp.data;

/**
 * Recursively pulls inner/nested blocks to return a flat array of blocks.
 *
 * @since 1.6.0
 *
 * @param {Array} blocks Array of WP Blocks.
 * @return {Array}
 */
export const flattenBlocks = ( blocks ) => {

	let flat = [];

	blocks.forEach( block => {
		if ( block.innerBlocks.length ) {
			flat = flat.concat( flattenBlocks( block.innerBlocks ) );
		} else {
			flat.push( block );
		}
	} );

	return flat;

};

/**
 * Retrieve an array of flattened blocks from the block editor.
 *
 * @since 1.6.0
 * @since 1.7.0 Backwards compat fix: fallback to `core/editor` if `core/block-editor` isn't available
 *
 * @return {Array}
 */
export default () => {

	const editor = select( 'core/block-editor') || select ( 'core/editor' );

	return flattenBlocks( editor.getBlocks() );

};
