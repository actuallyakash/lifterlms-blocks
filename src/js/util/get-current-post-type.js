/**
 * Utility to retrieve the current post type of the editor.
 *
 * This utility does not rely on waiting for full block editor initialization,
 * allowing primarily for us to conditionally register blocks based on
 * the current post type.
 *
 * @since 1.6.0
 * @version 1.6.0
 */

/**
 * Retrieve the current post type from LifterLMS script data.
 *
 * @since 1.5.0
 *
 * @return string|false
 */
export default () => {

	if ( window.llms && window.llms.post && window.llms.post.post_type ) {
		return window.llms.post.post_type;
	}

	return false;

};
