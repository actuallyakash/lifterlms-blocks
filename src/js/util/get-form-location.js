/**
 * Retrieve information about form locations
 *
 * @since [version]
 * @version [version]
 */

// External deps.
import { select } from '@wordpress/data';
import { store as editorStore } from '@wordpress/editor';

/**
 * Retrieve the form location id of the current form
 *
 * @since [version]
 *
 * @return {string} The ID of the current form's lcation
 */
export function getCurrentFormLocation() {
	const { getCurrentPost } = select( editorStore ),
		{ meta = {} } = getCurrentPost();

	return meta._llms_form_location || null;
}

/**
 * Retrieve the form location data object for the given form.
 *
 * @since [version]
 *
 * @return {?Object} Returns the location definition object or `null` if it doesn't exist.
 */
export function getCurrentFormLocationData() {
	return getFormLocationData( getCurrentFormLocation() );
}

/**
 * Retrieve form location data object for a give form by id
 *
 * @since [version]
 *
 * @param {string} id Form location ID.
 * @return {?Object} Returns the location definition object or `null` if it doesn't exist.
 */
export function getFormLocationData( id ) {
	const { formLocations } = window.llms;
	return formLocations[ id ] ? formLocations[ id ] : null;
}
