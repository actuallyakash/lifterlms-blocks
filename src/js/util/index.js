/**
 * Utilities
 *
 * @since 1.7.0
 * @version 1.7.0
 */

import getBlocksFlat, { flattenBlocks } from './get-blocks-flat';
import getCurrentPostType from './get-current-post-type.js';

export { getBlocksFlat, getCurrentPostType, flattenBlocks };
