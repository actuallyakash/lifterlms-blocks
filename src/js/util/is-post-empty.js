// WP Deps.
import { select } from '@wordpress/data';
import { store as editorStore } from '@wordpress/editor';

/**
 * Determine if the post content is completely empty
 *
 * @since [version]
 *
 * @return {boolean} [description]
 */
export default function () {
	const { getEditedPostContent } = select( editorStore ),
		content = getEditedPostContent();
	return ! content || ! content.includes( '<!-- wp:' );
}
