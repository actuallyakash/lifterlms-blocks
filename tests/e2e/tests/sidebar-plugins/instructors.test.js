/**
 * Test the instructors sidebar plugin
 *
 * @since 2.1.0
 * @version 2.1.0
 */

import { createNewPost } from "@wordpress/e2e-test-utils";
import { click, fillField } from "@lifterlms/llms-e2e-test-utils";

import { dragAndDrop, openSidebar } from "../../util";

const ITEM_SELECTOR = ".components-panel .llms-sortable-list .llms-instructor";

function getInstructorIndexSelector( nthChild ) {
	return `${ ITEM_SELECTOR }:nth-child( ${ nthChild } )`;
}

async function getInstructorNames() {
	return await page.$$eval(
		`${ ITEM_SELECTOR } .llms-instructor--header strong`,
		(els) => els.map(({ innerHTML }) => innerHTML)
	);
}

async function toggleEdit( nthChild ) {
	const INDEX_SELECTOR = getInstructorIndexSelector( nthChild );
	await click(`${ INDEX_SELECTOR } button[aria-label="Edit instructor"]`);
	return page.waitForTimeout( 500 );
}

async function reorderInstructor( draggingNthChild, droppingNthChild ) {

	const BUTTON_SELECTOR = 'button[aria-label="Reorder instructor"]';

	await dragAndDrop(
		`${ getInstructorIndexSelector( draggingNthChild ) } ${ BUTTON_SELECTOR }`,
		`${ getInstructorIndexSelector( droppingNthChild ) } ${ BUTTON_SELECTOR }` );
	await page.waitForTimeout( 500 );

}

describe( 'Sidebar/Plugins/Instructors', () => {

	beforeAll( async () => {

		page.once( 'dialog', async dialog => await dialog.accept() )

		const CLOSE_SELECTOR =
			'.editor-post-publish-panel button[aria-label="Close panel"]';

		await createNewPost( { postType: 'course', title: 'instructors' } );

		await openSidebar();

	} );

	test( 'Current user is the default primary instructor', async () => {

		// Admin user is the only instructor.
		expect( await getInstructorNames() ).toMatchSnapshot();

		// Primary instructor star exists.
		expect( await page.$eval( `${ ITEM_SELECTOR } .dashicons-star-filled`, ( el ) => el.outerHTML ? true : false ) ).toBe( true );

	} );

	test( 'Primary instructors cannot be removed', async () => {
		await toggleEdit( 1 ); // Open
		expect( await page.evaluate( ( sel ) => document.querySelector( sel ), `${ getInstructorIndexSelector( 1 ) } .llms-instructor--settings button.is-destructive` ) ).toBe( null );
		await toggleEdit( 1 ); // Close
	} );

	test( 'Instructors can be added', async () => {

		await fillField( '.llms-search--user .llms-search__input input[type="text"]', 'Knox' );

		await page.waitForTimeout( 1000 );

		await page.waitForTimeoutSelector( '.llms-search--user .llms-search__menu' );
		await page.keyboard.press( 'Enter' );

		await page.waitForTimeoutSelector( getInstructorIndexSelector( 2 ) );

		expect( await getInstructorNames() ).toMatchSnapshot();

	} );

	test( 'Instructors can be reordered', async () => {
		await reorderInstructor( 2, 1 );
		expect( await getInstructorNames() ).toMatchSnapshot();
	} );

	test( 'Non-primary instructors can be removed', async () => {

		const BUTTON_SELECTOR = `${ getInstructorIndexSelector( 2 ) } button.is-destructive`;

		await toggleEdit( 2 );
		await page.waitForTimeout( BUTTON_SELECTOR );
		await click( BUTTON_SELECTOR );
		await page.waitForTimeout( 500 );

		expect( await getInstructorNames() ).toMatchSnapshot();

	} );

} );
