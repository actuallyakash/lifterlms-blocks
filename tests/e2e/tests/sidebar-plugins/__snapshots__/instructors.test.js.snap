// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar/Plugins/Instructors Current user is the default primary instructor 1`] = `
Array [
  "admin",
]
`;

exports[`Sidebar/Plugins/Instructors Instructors can be added 1`] = `
Array [
  "admin",
  "Knox Harrington",
]
`;

exports[`Sidebar/Plugins/Instructors Instructors can be reordered 1`] = `
Array [
  "Knox Harrington",
  "admin",
]
`;

exports[`Sidebar/Plugins/Instructors Non-primary instructors can be removed 1`] = `
Array [
  "Knox Harrington",
]
`;
